### PR TITLE
Add parser / writer and their relevant tests for VASP's POSCAR format

### PR DESCRIPTION
**Pull request by chazeon**

### CODE REVIEW

**Comment by LaurentRDC** [comment]
Thanks for this PR! I'll take a detailed look either tonight or tomorrow

**Comment by chazeon** [comment]
Great! Thank you!

**Comment by LaurentRDC** [comment]
The tests are failing for a reason which is unrelated to your PR. Don't worry about it. I tested locally and things are looking good.

File: crystals/crystal.py [hunk 1]
```diff
@@ -283,6 +283,26 @@ def from_ase(cls, atoms, **kwargs):
...
+    @classmethod
+    def from_poscar(cls, path, **kwargs):
+        """
+        Returns a Crystal object created from an ASE Atoms object.
+        Keyword arguments are passed to the class constructor.
+
+        Parameters
+        ----------
+        atoms : ase.Atoms
+            Atoms group.
```
Comment (LaurentRDC) [inline crystals/crystal.py]: Documentation needs to be updated here

File: crystals/crystal.py [hunk 2]
```diff
@@ -751,6 +771,24 @@ def to_ase(self, **kwargs):
...
+
+        Parameters
+        ----------
+        filename : path-like
+            Path to a file. If the file already exists, it will be overwritten.
+        kwargs: 
+            Keyword arguments are passed to `writers.write_poscar`.
+
+        """
+        write_poscar(self, filename, **kwargs)
```
Comment (LaurentRDC) [inline crystals/crystal.py]: If the `*args` aren't used, then you don't need to include them in the function signature

File: crystals/parsers.py [hunk 3]
```diff
@@ -1123,3 +1123,50 @@ def atoms(self):
...
+            self._lattice_vectors = np.array([next(f).split() for _ in range(3)]).astype(np.float64) * scaling_factor
+            self._atom_types = list(zip(
+                next(f).strip().split(),
+                map(int, next(f).strip().split()),
+            ))
+            self._atoms = []
+            flag = next(f)
+
+            if flag.startswith("S"):
+                raise NotImplementedError()
```
Comment (LaurentRDC) [inline crystals/parsers.py]: What is this? If you can't implement it right now, can you add a descriptive error message?

File: crystals/parsers.py [hunk 4]
```diff
@@ -1123,3 +1123,50 @@ def atoms(self):
...
+    def __init__(self, filename, **kwargs):
+        self.filename = filename
+
+        with open(filename, mode="r") as f:
+
+            next(f)
+        
+            scaling_factor = float(next(f))
+
+            self._lattice_vectors = np.array([next(f).split() for _ in range(3)]).astype(np.float64) * scaling_factor
```
Comment (LaurentRDC) [inline crystals/parsers.py]: I'm pretty sure that numpy 1.20 has deprecated the use of `np.float64` in favour of the built-in `float`.

File: crystals/tests/data/issue5_MgSiO3.cartesian.POSCAR [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+     1.385475278         3.236904860         6.053581715
+     3.059424639         1.427895427         0.400817811
+     2.303347111         2.713514328         1.613600016
+     2.141552925         1.951285839         4.840800285
+     0.080896825         4.283685684         4.840800285
+     4.364002705         0.381114572         1.613600016
+     2.222450018         0.000000000         3.227200031
+     0.000000000         2.332400084         3.227200031
+     2.222450018         0.000000000         0.000000000
+     0.000000000         2.332400084         0.000000000
```
Comment (LaurentRDC) [inline crystals/tests/data/issue5_MgSiO3.cartesian.POSCAR]: Can you add a comment to say where this file comes from? If it has been converted from another file, can you write that down?

File: crystals/tests/data/issue5_MgSiO3.direct.POSCAR [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+     0.311699986         0.693900005         0.937899984
+     0.688299987         0.306100020         0.062099933
+     0.518199980         0.581700015         0.250000000
+     0.481800020         0.418299985         0.750000037
+     0.018199920         0.918299934         0.750000037
+     0.981799966         0.081700085         0.250000000
+     0.500000000         0.000000000         0.500000000
+     0.000000000         0.500000000         0.500000000
+     0.500000000         0.000000000         0.000000000
+     0.000000000         0.500000000         0.000000000
```
Comment (LaurentRDC) [inline crystals/tests/data/issue5_MgSiO3.direct.POSCAR]: Can you add a comment to say where this file comes from? If it has been converted from another file, can you write that down?

File: crystals/writers.py [hunk 7]
```diff
@@ -179,3 +180,63 @@ def ase_atoms(crystal, **kwargs):
...
+            x, y, z = vec / scaling_factor
+            file.write(f"{x:10.6f}  {y:10.6f}  {z:10.6f}")
+            file.write("\n")
+
+        # Write atomic data row-by-row
+        # For easier human readability, atoms are sorted
+        # by element
+        atoms = sorted(crystal)
+        grouped = [(elem, list(atoms)) for elem, atoms in
+                   groupby(atoms, key=lambda atom: atom.element)]
```
Comment (LaurentRDC) [inline crystals/writers.py]: Isn't the name `atoms` used for two different things here?

**Review by LaurentRDC — CHANGES_REQUESTED** [review]
Just a few little things to check

File: crystals/parsers.py (hunk 4 above)
Comment (chazeon) [reply crystals/parsers.py]: Good to know! I will just use the built-in float instead.

**Comment by chazeon** [comment]
I hope I have got all those issues resolved.

File: crystals/tests/data/issue5_MgSiO3.cartesian.POSCAR (hunk 5 above)
Comment (chazeon) [reply crystals/tests/data/issue5_MgSiO3.cartesian.POSCAR]: These are converted from `issue5_MgSiO3.cif` with VESTA. I add this to the description at the first line of the file.

File: crystals/tests/data/issue5_MgSiO3.direct.POSCAR (hunk 6 above)
Comment (chazeon) [reply crystals/tests/data/issue5_MgSiO3.direct.POSCAR]: These are converted from `issue5_MgSiO3.cif` with VESTA. I add this to the description at the first line of the file.

File: crystals/crystal.py (hunk 2 above)
Comment (chazeon) [reply crystals/crystal.py]: Thanks, I have remove the `*args`.

**Comment by LaurentRDC** [comment]
Great, everything looks good. Thanks!

These changes should come in version 1.4.0, released I would imagine in the next few days.

**Comment by LaurentRDC** [comment]
[Version 1.4.0](https://pypi.org/project/crystals/1.4.0/) is up with your contribution. Thank you so much!

**Pull request merged**